src/Symfony/Component/Routing/Matcher/Dumper/StaticPrefixCollection.php

```diff
@@ -0,0 +1,225 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Routing\Matcher\Dumper;
+
+/**
+ * Prefix tree of routes preserving routes order.
+ *
+ * @author Frank de Jonge <info@frankdejonge.nl>
+ *
+ * @internal
+ */
+class StaticPrefixCollection
+{
+    /**
+     * @var string
+     */
+    private $prefix;
+
+    /**
+     * @var array[]|StaticPrefixCollection[]
+     */
+    private $items = array();
+
+    /**
+     * @var int
+     */
+    private $matchStart = 0;
+
+    /**
+     * @param string $prefix
+     */
+    public function __construct($prefix = '')
+    {
+        $this->prefix = $prefix;
+    }
+
+    /**
+     * @return string
+     */
+    public function getPrefix()
+    {
+        return $this->prefix;
+    }
+
+    /**
+     * @return mixed[]|StaticPrefixCollection[]
+     */
+    public function getItems()
+    {
+        return $this->items;
+    }
+
+    /**
+     * Adds a route to a group.
+     *
+     * @param string $prefix
+     * @param mixed  $route
+     */
+    public function addRoute($prefix, $route)
+    {
+        $this->guardAgainstAddingNotAcceptedRoutes($prefix);
+
+        if ($this->prefix === $prefix) {
+            // When a prefix is exactly the same as the base we move up the match start position.
+            // This is needed because otherwise routes that come afterwards have higher precedence
+            // than a possible regular expression, which goes against the input order sorting.
+            $this->items[] = array($prefix, $route);
+            $this->matchStart = count($this->items);
+
+            return;
+        }
+
+        foreach ($this->items as $i => $item) {
+            if ($i < $this->matchStart) {
+                continue;
+            }
+
+            if ($item instanceof self && $item->accepts($prefix)) {
+                $item->addRoute($prefix, $route);
+
+                return;
+            }
+
+            $group = $this->groupWithItem($item, $prefix, $route);
+
+            if ($group instanceof self) {
+                $this->items[$i] = $group;
+
+                return;
+            }
+        }
+
+        // No optimised case was found, in this case we simple add the route for possible
+        // grouping when new routes are added.
+        $this->items[] = array($prefix, $route);
+    }
+
+    /**
+     * Tries to combine a route with another route or group.
+     *
+     * @param StaticPrefixCollection|array $item
+     * @param $prefix
+     * @param $route
+     *
+     * @return bool|StaticPrefixCollection
```

Comment:
please use ``null``, not ``false`` for the not found case. This would create an API with a nullable return type rather than with a mixed type